### PR TITLE
Switched Hawkular Metrics to use certificates generated by OpenShift

### PR DESCRIPTION
**Pull request by jpkrohling**

### CODE REVIEW

**Comment by jpkrohling** [comment]
This builds on top of the previous PRs, #295 , #296 and #297 and will be rebased accordingly once those are merged.


This doesn't work with Heapster yet: right after preparing this PR, I've noticed the following on Heapster's logs:

```
E0210 11:14:35.352224       1 client.go:192] Could not update tags: Hawkular returned status code 403, error message: Reply could not be parsed: invalid character '<' looking for beginning of value
```

So, please *don't merge* this yet. Other than that, this PR is ready to be reviewed. As it is, with the dependent commits added, the following command can be used at the root of this repository to create the whole infra, with 2 Cassandra nodes:

```
oc process -f metrics.yaml \
   -v IMAGE_PREFIX="jpkroehling/origin-" \
   -v IMAGE_VERSION="dev" \
   -v HAWKULAR_METRICS_HOSTNAME=hawkular-metrics.example.com \
   -v USE_PERSISTENT_STORAGE=false \
   -v CASSANDRA_NODES=2 | oc create -n openshift-infra -f -
```

**Comment by jcantrill** [comment]
[test]

File: deployer/scripts/hawkular.sh [hunk 1]
```diff
@@ -121,34 +57,10 @@ EOF
...
-   if [ "$type" = "reencrypt" ]; then
-      oc patch route hawkular-metrics -p "$(python <<-EOF
-				import json
-				ca = open("$dir/hawkular-metrics-ca.cert").read()
-				print json.dumps({"spec": {"tls": {"destinationCACertificate": ca}}})
-				EOF
-      )"
-   fi
-  fi
-
```
Comment (mwringe) [inline deployer/scripts/hawkular.sh]: Can we actually remove this section. When doing a fresh we want the route's destinationCACertificate to match the ca certificate used to sign the Hawkular Metrics service

File: deployer/scripts/hawkular.sh [hunk 2]
```diff
@@ -121,34 +57,10 @@ EOF
...
-    `openssl crl2pkcs7 -nocrl -certfile ${secret_dir}/hawkular-metrics.pem | openssl pkcs7 -print_certs | grep -v "^subject=*\|^issuer=*\|^$" > $dir/custom-certificate.crt`
-     route_params="${route_params} --cert=$dir/custom-certificate.crt --key=$dir/custom-certificate.key"
-     if [ -s ${secret_dir}/hawkular-metrics-ca.cert ]; then
-       route_params="${route_params} --ca-cert=${secret_dir}/hawkular-metrics-ca.cert"
-     fi
-  fi
   # this may return an error code if the route already exists, this is to be expect with a refresh and is why we have the || true here
-  oc create route reencrypt hawkular-metrics ${route_params} || true
+  ## once BZ 1401081 is done, the Route specified on `hawkular-metrics.yaml` should work and this command here should be removed.
+  oc create route passthrough --service hawkular-metrics || true
```
Comment (mwringe) [inline deployer/scripts/hawkular.sh]: We don't want a passthrough route, we need a re-encrypt route

File: deployer/templates/hawkular-metrics.yaml [hunk 3]
```diff
@@ -134,9 +130,18 @@ objects:
...
+#  metadata:
+#    creationTimestamp: null
+#    name: hawkular-metrics-route
+#  spec:
+#    tls:
+#      insecureEdgeTerminationPolicy: Allow
+#      termination: reencrypt
+#    to:
+#      kind: Service
+#      name: hawkular-metrics
```
Comment (mwringe) [inline deployer/templates/hawkular-metrics.yaml]: Why are we removing the route definition here?

Comment (mwringe) [inline deployer/templates/hawkular-metrics.yaml]: Why do we have a commented out section here?

File: deployer/templates/heapster.yaml [hunk 4]
```diff
@@ -83,9 +82,9 @@ objects:
           - "--wrapper.username_file=/hawkular-account/hawkular-metrics.username"
           - "--wrapper.password_file=/hawkular-account/hawkular-metrics.password"
           - "--wrapper.allowed_users_file=/secrets/heapster.allowed-users"
-          - "--wrapper.endpoint_check=https://hawkular-metrics:443/hawkular/metrics/status"
+          - "--wrapper.endpoint_check=https://hawkular-metrics.openshift-infra.svc.cluster.local:443/hawkular/metrics/status"
```
Comment (mwringe) [inline deployer/templates/heapster.yaml]: This is now hardcoding the project to 'openshift-infra'. You can either change this back to the way it was, or even better, get the namespace using the downward api and replace 'openshift-infra' with the actual namespace value

File: hawkular-metrics/hawkular-metrics-wrapper.sh [hunk 5]
```diff
@@ -91,16 +91,46 @@ fi
...
 
 # Setup the truststore so that it will accept the OpenShift cert
+KEYTOOL_COMMAND=/usr/lib/jvm/java-1.8.0/jre/bin/keytool
 HAWKULAR_METRICS_AUTH_DIR=$HAWKULAR_METRICS_DIRECTORY/auth
 mkdir $HAWKULAR_METRICS_AUTH_DIR
 pushd $HAWKULAR_METRICS_AUTH_DIR
 
-cp $KEYSTORE hawkular-metrics.keystore
-cp $TRUSTSTORE hawkular-metrics.truststore
+if [ ! -n "$KEYSTORE" ]; then
```
Comment (mwringe) [inline hawkular-metrics/hawkular-metrics-wrapper.sh]: Is there any reason why we should be checking for this? I think it makes sense to just require that the certificate and keys are passed in, and don't give the option of passing in a keystore

File: hawkular-metrics/hawkular-metrics-wrapper.sh [hunk 6]
```diff
@@ -91,16 +91,46 @@ fi
...
+    ${KEYTOOL_COMMAND} -v -importkeystore -srckeystore ${PKCS12_FILE} -srcstoretype PKCS12 -destkeystore ${KEYSTORE} -deststoretype JKS -deststorepass ${KEYSTORE_PASSWORD} -srcstorepass ${KEYSTORE_PASSWORD}
+    if [ $? != 0 ]; then
+        echo "Failed to create a Java Keystore file with the service-specific certificate. Aborting."
+        exit 1
+    fi
+else
+    cp $KEYSTORE hawkular-metrics.keystore
+fi
+
+if [ -n "$TRUSTSTORE" ]; then
```
Comment (mwringe) [inline hawkular-metrics/hawkular-metrics-wrapper.sh]: We should be generating the truststore and not give a user an option to pass one in

File: deployer/scripts/hawkular.sh (hunk 1 above)
Comment (jpkrohling) [inline deployer/scripts/hawkular.sh]: A cert is created once via a `secret`. Similarly, a `service-ca` is created once for the lifecycle of a cluster. So, even if the cert's secret is removed, it will still be signed by the same CA.

File: deployer/scripts/hawkular.sh (hunk 2 above)
Comment (jpkrohling) [inline deployer/scripts/hawkular.sh]: Right, hence the comment on line 61. Once BZ 1401081 is done, we'll be able to create the route and specify the generated cert (which is also why the route specification is commented out on the YAML file).

File: deployer/templates/hawkular-metrics.yaml (hunk 3 above)
Comment (jpkrohling) [inline deployer/templates/hawkular-metrics.yaml]: Because this still doesn't work (BZ 1401081) :/

File: deployer/templates/heapster.yaml (hunk 4 above)
Comment (jpkrohling) [inline deployer/templates/heapster.yaml]: Changing it back isn't an option, as the name on the cert is a FQDN, not just `hawkular-metrics` (IIRC). But I can certainly make it use the API. I just need to figure out which API you mean :)

File: hawkular-metrics/hawkular-metrics-wrapper.sh (hunk 5 above)
Comment (jpkrohling) [inline hawkular-metrics/hawkular-metrics-wrapper.sh]: The keystore can be specified via `--hmw.keystore`. If it's not specified, we load the certs from the location we expect them to be. I'm not familiar with the possible use cases, so, not sure if it's useful, but my idea would be that users would be able to override the whole keystore and truststore, in case they need more than one cert on the keystore.

**Comment by jpkrohling** [comment]
Status: this is rebased, but a couple of the comments from the last review weren't fixed yet.

**Comment by jpkrohling** [comment]
@mwringe , would you please review it again? If this looks OK , I'll apply similar changes to the playbook.

File: deployer/run.sh [hunk 7]
```diff
@@ -95,13 +95,12 @@ rm -rf $dir && mkdir -p $dir && chmod 700 $dir || :
 mkdir -p $secret_dir && chmod 700 $secret_dir || :
 
 hawkular_metrics_hostname=${HAWKULAR_METRICS_HOSTNAME:-hawkular-metrics.example.com}
-hawkular_metrics_alias=${HAWKULAR_METRICS_ALIAS:-hawkular-metrics}
 
 openshift admin ca create-signer-cert  \
   --key="${dir}/ca.key" \
   --cert="${dir}/ca.crt" \
   --serial="${dir}/ca.serial.txt" \
-  --name="metrics-signer@$(date +%s)"
```
Comment (jpkrohling) [inline deployer/run.sh]: Revert this change.

File: deployer/run.sh [hunk 8]
```diff
@@ -83,7 +83,7 @@ metric_resolution=${METRIC_RESOLUTION:-30s}
 project=${PROJECT:-openshift-infra}
 
 # the master certificate and service account tokens
-master_ca=${MASTER_CA:-/var/run/secrets/kubernetes.io/serviceaccount/ca.crt}
+master_ca=${MASTER_CA:-/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt}
```
Comment (jpkrohling) [inline deployer/run.sh]: Confirm with @mwringe if this is correct. IIRC, he mentioned that this might not be appropriate.

File: deployer/templates/heapster.yaml (hunk 4 above)
Comment (mwringe) [inline deployer/templates/heapster.yaml]: The downward api can be used to request information, such as which namespace your project is deployed tohttps://kubernetes.io/docs/tasks/configure-pod-container/downward-api-volume-expose-pod-information/

See: https://github.com/openshift/origin-metrics/blob/master/deployer/templates/hawkular-metrics.yaml#L106-L109

You just need to add that to your template and use the envar here

File: deployer/run.sh (hunk 8 above)
Comment (mwringe) [inline deployer/run.sh]: This is for connecting to the master api, which is signed with the main ca certificate and not the service ca certificate.

**Comment by mwringe** [comment]
I think I have mentioned all my main concerns here, once those are fixed and I can verify that it works then it should be good to be merged.

File: deployer/templates/heapster.yaml (hunk 4 above)
Comment (jpkrohling) [inline deployer/templates/heapster.yaml]: Cool, didn't know it had a name :)

**Comment by jpkrohling** [comment]
I have just added a new commit to this PR, with a bug fix and with the changes requested from reviews.

Would you please check this last commit? If that looks good, I'll squash and update this PR, as well as apply the changes to the playbook.

**Comment by mwringe** [comment]
[merge]
